docs/howto_config.md
# Configuration
The MicroShift configuration file must be located at `/etc/microshift/config.yaml`. A sample `/etc/microshift/config.yaml.default` configuration file is installed by the MicroShift RPM and it can be used as a template when customizing MicroShift.

The format of the `config.yaml` configuration file is as follows.
<!---
{{- template "docsReplaceBasic" . }}
{{- with deleteCurrent -}}
--->
```yaml
apiServer:
    advertiseAddress: ""
    subjectAltNames:
        - ""
debugging:
    logLevel: ""
dns:
    baseDomain: ""
etcd:
    memoryLimitMB: 0
manifests:
    kustomizePaths:
        - ""
network:
    clusterNetwork:
        - ""
    serviceNetwork:
        - ""
    serviceNodePortRange: ""
node:
    hostnameOverride: ""
    nodeIP: ""

```
<!---
{{- end }}
--->

## Default Settings

In case `config.yaml` is not provided, the following default settings will be used.
<!---
{{- template "docsReplaceFull" . }}
{{- with deleteCurrent -}}
--->
```yaml
apiServer:
    advertiseAddress: ""
    subjectAltNames:
        - ""
debugging:
    logLevel: Normal
dns:
    baseDomain: example.com
etcd:
    memoryLimitMB: 0
manifests:
    kustomizePaths:
        - /usr/lib/microshift/manifests
        - /usr/lib/microshift/manifests.d/*
        - /etc/microshift/manifests
        - /etc/microshift/manifests.d/*
network:
    clusterNetwork:
        - 10.42.0.0/16
    serviceNetwork:
        - 10.43.0.0/16
    serviceNodePortRange: 30000-32767
node:
    hostnameOverride: ""
    nodeIP: ""

```
<!---
{{- end }}
--->

## Service NodePort range

The `serviceNodePortRange` setting allows the extension of the port range available
to NodePort Services. This option is useful when specific standard ports under the
`30000-32767` need to be exposed. i.e., your device needs to expose the `1883/tcp`
MQTT port on the network because some client devices cannot use a different port.

If you use this option, you must be careful; NodePorts can overlap with system ports,
causing malfunction of the system or MicroShift. Take the following considerations
into account:

* Do not create any NodePort service without an explicit `nodePort` selection, in this
  case the port is assigned randomly by the kube-apiserver.
* Do not create any NodePort service for any system service port, MicroShift port,
  or other services you expose on your device HostNetwork.

List of ports that you must avoid:

| Port          | Description                                                     |
|---------------|-----------------------------------------------------------------|
| 22/tcp        | SSH port
| 80/tcp        | OpenShift Router HTTP endpoint
| 443/tcp       | OpenShift Router HTTPS endpoint
| 1936/tcp      | Metrics service for the openshift-router, not exposed today
| 2379/tcp      | etcd port
| 2380/tcp      | etcd port
| 6443          | kubernetes API
| 8445/tcp      | openshift-route-controller-manager
| 9537/tcp      | cri-o metrics
| 10250/tcp     | kubelet
| 10248/tcp     | kubelet healthz port
| 10259/tcp     | kube scheduler
|---------------|-----------------------------------------------------------------|

## Etcd Memory Limit

By default, etcd will be allowed to use as much memory as it needs to handle the load on the system; however, in memory constrained systems, it may be preferred or necessary to limit the amount of memory etcd is allowed to use at a given time.

Setting the `memoryLimitMB` to a value greater than 0 will result in a soft memory limit being applied to etcd; etcd will be allowed to go over this value during operation, but memory will be more aggresively reclaimed from it if it does. A value of `128` megabytes is the  configuration floor - attempting to set the limit below 128 megabytes will result in the configuration being 128 megabytes.

Please note that values close to the floor may be more likely to impact etcd performance - the memory limit is a trade-off of memory footprint and etcd performance. The lower the limit, the more time etcd will spend on paging memory to disk and will take longer to respond to queries or even timing requests out if the limit is low and the etcd usage is high.

# Auto-applying Manifests

MicroShift leverages `kustomize` for Kubernetes-native templating and declarative management of resource objects. Upon start-up, it searches `/etc/microshift/manifests`, `/etc/microshift/manifests.d/*`, `/usr/lib/microshift/manifests`, and `/usr/lib/microshift/manifests.d/*` directories for a `kustomization.yaml`, `kustomization.yml`, or `Kustomization` file. If it finds one, it automatically runs `kubectl apply -k` command to apply that manifest.

Loading from multiple directories allows you to manage MicroShift workloads more flexibly. Different workloads can be independent of each other, instead of having to be loaded from a merged set of inputs.

| Location                          | Intent |
|-----------------------------------|--------|
| /etc/microshift/manifests         | Read-write location for configuration management systems or development
| /etc/microshift/manifests.d/*     | Read-write location for configuration management systems or development
| /usr/lib/microshift/manifests     | Read-only location for embedding configuration manifests on ostree based systems
| /usr/lib/microshift/manifestsd./* | Read-only location for embedding configuration manifests on ostree based systems

To override the list of paths, set `manifests.kustomizePaths` in the configuration file.

```yaml
manifests:
    kustomizePaths:
        - "/opt/alternate/path"
```

The values of `kustomizePaths` may be glob patterns.

```yaml
manifests:
    kustomizePaths:
        - "/opt/alternative/path.d/*"
```

To disable loading manifests, set the configuration option to an empty
list.

```yaml
manifests:
    kustomizePaths: []
```


## Manifest Example

The example demonstrates automatic deployment of a `busybox` container using `kustomize` manifests in the `/etc/microshift/manifests` directory.

Run the following command to create the manifest files.

```bash
MANIFEST_DIR=/etc/microshift/manifests
sudo mkdir -p ${MANIFEST_DIR}

sudo tee ${MANIFEST_DIR}/busybox.yaml &>/dev/null <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: busybox
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: busybox-deployment
spec:
  selector:
    matchLabels:
      app: busybox
  template:
    metadata:
      labels:
        app: busybox
    spec:
      containers:
      - name: busybox
        image: BUSYBOX_IMAGE
        command:
          - sleep
          - "3600"
EOF

sudo tee ${MANIFEST_DIR}/kustomization.yaml &>/dev/null <<EOF
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: busybox
resources:
  - busybox.yaml
images:
  - name: BUSYBOX_IMAGE
    newName: busybox:1.35
EOF
```

Restart the MicroShift service to apply the manifests and verify that the `busybox` pod is running.

```bash
sudo systemctl restart microshift
oc get pods -n busybox
```

## Storage Configuration

MicroShift's included CSI plugin manages LVM LogicialVolumes to provide persistent workload storage. For specific storage configuration, refer to the dedicated [documentation](./storage/configuration.md).


pkg/config/config.go
package config

//go:generate ../../scripts/generate-config.sh

import (
	"bytes"
	"fmt"
	"net"
	"net/url"
	"os"
	"os/exec"
	"strings"
	"time"

	"k8s.io/apimachinery/pkg/util/sets"
	"k8s.io/klog/v2"

	"github.com/apparentlymart/go-cidr/cidr"
	"github.com/openshift/microshift/pkg/util"
)

const (
	// default DNS resolve file when systemd-resolved is used
	DefaultSystemdResolvedFile = "/run/systemd/resolve/resolv.conf"
)

type Config struct {
	DNS       DNS        `json:"dns"`
	Network   Network    `json:"network"`
	Node      Node       `json:"node"`
	ApiServer ApiServer  `json:"apiServer"`
	Etcd      EtcdConfig `json:"etcd"`
	Debugging Debugging  `json:"debugging"`
	Manifests Manifests  `json:"manifests"`

	// Internal-only fields
	Ingress      IngressConfig `json:"-"`
	userSettings *Config       `json:"-"` // the values read from the config file

	MultiNode MultiNodeConfig `json:"-"` // the value read from commond line
}

// NewDefault creates a new Config struct populated with the
// default values and with any computed values updated based on those
// defaults.
func NewDefault() *Config {
	c := &Config{}
	if err := c.fillDefaults(); err != nil {
		klog.Fatalf("Failed to initialize config: %v", err)
	}
	if err := c.updateComputedValues(); err != nil {
		klog.Fatalf("Failed to initialize config: %v", err)
	}
	return c
}

// fillDefaults forcibly sets the configuration to the default
// values. We do not use a static struct for the defaults because some
// of them are computed from the environment. If any error occurs
// probing the environment, the values in the Config instance are not
// changed.
func (c *Config) fillDefaults() error {
	// Look up any values that may generate an error
	subjectAltNames, err := getAllHostnames()
	if err != nil {
		return fmt.Errorf("failed to get all hostnames: %v", err)
	}
	hostname, err := os.Hostname()
	if err != nil {
		return fmt.Errorf("Failed to get hostname %v", err)
	}
	nodeIP, err := util.GetHostIP()
	if err != nil {
		return fmt.Errorf("failed to get host IP: %v", err)
	}

	c.Debugging = Debugging{
		LogLevel: "Normal",
	}
	c.ApiServer = ApiServer{
		SubjectAltNames: subjectAltNames,
		URL:             "https://localhost:6443",
		Port:            6443,
	}
	c.Node = Node{
		HostnameOverride: hostname,
		NodeIP:           nodeIP,
	}
	c.DNS = DNS{
		BaseDomain: "example.com",
	}
	c.Network = Network{
		ClusterNetwork: []string{
			"10.42.0.0/16",
		},
		ServiceNetwork: []string{
			"10.43.0.0/16",
		},
		ServiceNodePortRange: "30000-32767",
		DNS:                  "10.43.0.10",
	}
	c.Etcd = EtcdConfig{
		MemoryLimitMB:           0,
		QuotaBackendBytes:       8 * 1024 * 1024 * 1024,
		MinDefragBytes:          100 * 1024 * 1024,
		MaxFragmentedPercentage: 45,
		DefragCheckFreq:         5 * time.Minute,
	}
	c.Manifests = Manifests{
		KustomizePaths: []string{defaultManifestDirLib, defaultManifestDirEtc},
	}

	c.MultiNode.Enabled = false

	return nil
}

// incorporateUserSettings merges any values read from the
// configuration file provided by the user with the existing settings
// (usually the defaults).
func (c *Config) incorporateUserSettings(u *Config) {
	c.userSettings = u

	if u.DNS.BaseDomain != "" {
		c.DNS.BaseDomain = u.DNS.BaseDomain
	}

	if len(u.Network.ClusterNetwork) != 0 {
		c.Network.ClusterNetwork = u.Network.ClusterNetwork
	}
	if len(u.Network.ServiceNetwork) != 0 {
		c.Network.ServiceNetwork = u.Network.ServiceNetwork
		// The default for the API server address is computed from the
		// service network. If the user provides a network without
		// also overriding the computed address, we need to clear the
		// address here so it is recomputed later. If they provide
		// both the network and the address, the address will be
		// copied into place below with the other API server settings.
		if u.ApiServer.AdvertiseAddress == "" {
			c.ApiServer.AdvertiseAddress = ""
		}
	}
	if u.Network.ServiceNodePortRange != "" {
		c.Network.ServiceNodePortRange = u.Network.ServiceNodePortRange
	}
	if u.Network.DNS != "" {
		c.Network.DNS = u.Network.DNS
	}

	if u.Etcd.MemoryLimitMB != 0 {
		c.Etcd.MemoryLimitMB = u.Etcd.MemoryLimitMB
	}

	if u.Node.HostnameOverride != "" {
		c.Node.HostnameOverride = u.Node.HostnameOverride
	}
	if u.Node.NodeIP != "" {
		c.Node.NodeIP = u.Node.NodeIP
	}
	if len(u.ApiServer.SubjectAltNames) != 0 {
		c.ApiServer.SubjectAltNames = u.ApiServer.SubjectAltNames
	}
	if u.ApiServer.AdvertiseAddress != "" {
		c.ApiServer.AdvertiseAddress = u.ApiServer.AdvertiseAddress
	}
	if u.ApiServer.URL != "" {
		c.ApiServer.URL = u.ApiServer.URL
	}

	if u.Debugging.LogLevel != "" {
		c.Debugging.LogLevel = u.Debugging.LogLevel
	}

	// Check for nil instead of an empty list because if a user
	// provides a list but it is empty we want to treat that as
	// disabling the manifest loader.
	if u.Manifests.KustomizePaths != nil {
		c.Manifests.KustomizePaths = u.Manifests.KustomizePaths
	}
}

// updateComputedValues examins the existing settings and converts any
// inputs to more easily consumable units or fills in any defaults
// computed based on the values of other settings.
func (c *Config) updateComputedValues() error {
	clusterDNS, err := c.computeClusterDNS()
	if err != nil {
		return err
	}
	c.Network.DNS = clusterDNS

	// If KAS advertise address configured, we do not want to apply
	// the IP to the internal interface.
	if c.userSettings != nil && len(c.userSettings.ApiServer.AdvertiseAddress) != 0 {
		c.ApiServer.SkipInterface = true
	}

	// If we have no advertise address, pick one.
	if len(c.ApiServer.AdvertiseAddress) == 0 {
		// unchecked error because this was done when getting cluster DNS
		_, svcNet, _ := net.ParseCIDR(c.Network.ServiceNetwork[0])
		// Since the KAS advertise address was not provided we will default to the
		// next immediate subnet after the service CIDR. This is due to the fact
		// that using the actual apiserver service IP as an endpoint slice breaks
		// host network pods trying to reach apiserver, as the VIP 10.43.0.1:443 is
		// not translated to 10.43.0.1:6443. It remains unchanged and therefore
		// connects to the ingress router instead, triggering all sorts of errors.
		nextSubnet, exceed := cidr.NextSubnet(svcNet, 32)
		if exceed {
			return fmt.Errorf("unable to compute next subnet from service CIDR")
		}
		// First and last are the same because of the /32 netmask.
		firstValidIP, _ := cidr.AddressRange(nextSubnet)
		c.ApiServer.AdvertiseAddress = firstValidIP.String()
	}

	return nil
}

func (c *Config) validate() error {
	//nolint:nestif // extracting the nested ifs will just increase the complexity of the if expressions as validation expands
	if len(c.ApiServer.SubjectAltNames) > 0 {
		// Any entry in SubjectAltNames will be included in the external access certificates.
		// Any of the hostnames and IPs (except the node IP) listed below conflicts with
		// other certificates, such as the service network and localhost access.
		// The node IP is a bit special. Apiserver k8s service, which holds a service IP
		// gets resolved to the node IP. If we include the node IP in the SAN then we have
		// an ambiguity, the same IP matches two different certificates and there are errors
		// when trying to reach apiserver from within the cluster using the service IP.
		// Apiserver will decide which certificate to return to client hello based on SNI
		// (which client-go does not use) or raw IP mappings. As soon as there is a match for
		// the node IP it returns that certificate, which is the external access one. This
		// breaks all pods trying to reach apiserver, as hostnames dont match and the certificate
		// is invalid.
		u, err := url.Parse(c.ApiServer.URL)
		if err != nil {
			return fmt.Errorf("failed to parse cluster URL: %v", err)
		}
		if u.Hostname() == "localhost" || u.Hostname() == "127.0.0.1" {
			if stringSliceContains(c.ApiServer.SubjectAltNames, "localhost", "127.0.0.1") {
				return fmt.Errorf("subjectAltNames must not contain localhost, 127.0.0.1")
			}
		} else {
			if stringSliceContains(c.ApiServer.SubjectAltNames, c.Node.NodeIP) {
				return fmt.Errorf("subjectAltNames must not contain node IP")
			}
			if !stringSliceContains(c.ApiServer.SubjectAltNames, u.Host) || u.Host != c.Node.HostnameOverride {
				return fmt.Errorf("Cluster URL host %v is not included in subjectAltNames or nodeName", u.String())
			}
		}
		if stringSliceContains(
			c.ApiServer.SubjectAltNames,
			"kubernetes",
			"kubernetes.default",
			"kubernetes.default.svc",
			"kubernetes.default.svc.cluster.local",
			"openshift",
			"openshift.default",
			"openshift.default.svc",
			"openshift.default.svc.cluster.local",
			c.ApiServer.AdvertiseAddress,
		) {
			return fmt.Errorf("subjectAltNames must not contain apiserver kubernetes service names or IPs")
		}
	}

	if c.Etcd.MemoryLimitMB > 0 && c.Etcd.MemoryLimitMB < EtcdMinimumMemoryLimit {
		return fmt.Errorf("etcd.memoryLimitMB value %d is below the minimum allowed %d",
			c.Etcd.MemoryLimitMB, EtcdMinimumMemoryLimit,
		)
	}

	return nil
}

var allHostnames []string

func getAllHostnames() ([]string, error) {
	if len(allHostnames) != 0 {
		return allHostnames, nil
	}
	cmd := exec.Command("/bin/hostname", "-A")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		return nil, fmt.Errorf("Error when executing 'hostname -A': %v", err)
	}
	outString := out.String()
	outString = strings.Trim(outString[:len(outString)-1], " ")
	// Remove duplicates to avoid having them in the certificates.
	names := strings.Split(outString, " ")
	set := sets.NewString(names...)
	allHostnames = set.List()
	return allHostnames, nil
}
